File: src/main/java/org/usfirst/frc/team1458/lib/pid/PIDConstants.java
package org.usfirst.frc.team1458.lib.pid;


/**
 * Container for PID constants
 * @author asinghani
 */
public class PIDConstants {
	/**
	 * Proportional
	 */
	public final double kP;

	/**
	 * Integral
	 */
	public final double kI;

	/**
	 * Derivative
	 */
	public final double kD;

	/**
	 * Feedforward
	 */
	public final double kF;

	public static final PIDConstants ZERO = new PIDConstants(0, 0, 0, 0);

	/**
	 * @param kP
	 */
	public PIDConstants(double kP) {
		this(kP, 0.0, 0.0, 0.0);
	}

	/**
	 * @param kP
	 * @param kI
	 * @param kD
	 */
	public PIDConstants(double kP, double kI, double kD) {
		this(kP, kI, kD, 0.0);
	}

	/**
	 * @param kP
	 * @param kI
	 * @param kD
	 * @param kF
	 */
	public PIDConstants(double kP, double kI, double kD, double kF) {
		this.kP = kP;
		this.kI = kI;
		this.kD = kD;
		this.kF = kF;
	}
}

File: src/main/java/org/usfirst/frc/team1458/lib/pid/PIDConstants.kt
package org.usfirst.frc.team1458.lib.pid

/**
 * Class to hold PID constants
 *
 * @param kP Proportional gain
 * @param kI Integral gain
 * @param kD Derivative gain
 * @param kF Feedforward gain
 * @param kF_func Custom feedforward function (allows the feedforward to be calculated from multiple inputs)
 */
data class PIDConstants(val kP: Double, val kI: Double = 0.0, val kD: Double = 0.0, val kF: Double = 0.0, val kF_func: (DoubleArray) -> Double = { 0.0 }) {
    companion object {
        val DISABLE = PIDConstants(0.0, 0.0, 0.0, 0.0, { 0.0 })
    }
}